### PR TITLE
Private/deliz/ca 332753

### PR DESCRIPTION
**Pull request by DeliZhangX**

### CODE REVIEW

File: autocertkit/utils.py [hunk 1]
```diff
@@ -1166,74 +1166,123 @@ def get_context_test_ifs(vm_ref):
...
+        # their own interfaces (avoid interface forwarding).
+        call_ack_plugin(session, 'reset_arp', {'vm_ref': vm_ref, 'mip': mip})
 
-    if not device.startswith('eth'):
-        raise Exception(
-            "Invalid device specified, it should be in the format 'ethX'")
 
-    start = time.time()
+def disable_vm_static_ip_service(session, mip):
+    cmd = """systemctl stop static-ip; systemctl disable static-ip; systemctl status static-ip; """ \
```
Comment (acefei) [inline autocertkit/utils.py]: Given using multiline string, so you might remove the redundant `""" \ """`

Comment (DeliZhangX) [reply autocertkit/utils.py]: This is to avoid indent checking

Comment (acefei) [reply autocertkit/utils.py]: > This is to avoid indent checking

 I think `textwrap.dedent(cmd).` is what you want.

File: autocertkit/utils.py [hunk 2]
```diff
@@ -1166,74 +1166,123 @@ def get_context_test_ifs(vm_ref):
...
-        log.debug("VM %s has these %s IPs" % (vm_ref, ips))
+def get_vm_vif_ifs(session, vm_ref):
+    """Get vm interfaces information (vif, mac, ip) from xenstore vif records"""
+    ifs = {}
+    dom_root = "/local/domain/%s" % str(session.xenapi.VM.get_domid(vm_ref))
 
-        for k, v in ips.iteritems():
-            if k == device:
-                return v
+    cmd = b'''xenstore-ls -f %s | grep -E "%s/device/vif/[0-9]+/mac|%s/attr/vif/[0-9]+/ipv4/[0-9]+"''' % (dom_root, dom_root, dom_root)
```
Comment (acefei) [inline autocertkit/utils.py]: Using `str.format()` can aviod the redundant `dom_root` paramaters.

File: plugins/autocertkit [hunk 3]
```diff
@@ -605,6 +557,7 @@ def deploy_vf_driver(session, args):
     """Install VF driver on a specified VM"""
     log.debug("deploy VF driver plugin call")
     vm_ref = validate_exists(args, 'vm_ref')
+    mip = validate_exists(args, 'mip')
```
Comment (xihuan-citrix) [inline plugins/autocertkit]: why not do the same as in deploy_iperf, move to `if not session.xenapi.VM.get_is_control_domain(vm_ref):`

File: autocertkit/cpu_tests.py [hunk 4]
```diff
@@ -78,13 +79,8 @@ def _create_test_threads(self, session, vm_ref_list):
...
         threads = []
         for vm_ref in vm_ref_list:
-            vm_ip = wait_for_ip(session, vm_ref, 'eth0')
-            threads.append(create_test_thread(lambda: TimeoutFunction(ssh_command(vm_ip,
-                                                                                  self.username,
-                                                                                  self.password,
-                                                                                  self.cmd_str,
-                                                                                  timeout=self.timeout)["stdout"],
-                                                                      self.timeout, '%s test timed out %d' % (self.test, self.timeout))))
+            threads.append(create_test_thread(ssh_command,
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/cpu_tests.py]: I never see this here timeout-ed.
Will revert to before.

File: autocertkit/network_tests.py [hunk 5]
```diff
@@ -1353,30 +1231,23 @@ def deploy_droid_vms(self, session, vf_driver, network_refs, sms):
...
-        if self.vf_num_test < 16:
-            test_times = 8
-        elif self.vf_num_test >= 16 and self.vf_num_test < 32:
-            test_times = 4
-        elif self.vf_num_test >= 32 and self.vf_num_test < 64:
-            test_times = 2
-        else:
-            test_times = 1
+        test_times = 10
+        test_timeout = 3600
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/network_tests.py]: It should use the new way before, vf_num is a incorrect way.

File: autocertkit/storage_tests.py [hunk 6]
```diff
@@ -63,20 +63,21 @@ def _setup_vms(self, session):
...
-                                              sr_ref)
+        return deploy_common_droid_vms_on_hosts(session,
+                                                [host_ref],
+                                                [net_ref],
+                                                self.vm_count,
+                                                {net_ref: self.get_static_manager(
+                                                    net_ref)},
+                                                sr_ref)[host_ref]
 
     def _call_plugin(self, session, vm_ref_list, call):
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/storage_tests.py]: Yes.
I don't/shouldn't want to change anything unrelated to the commit

File: autocertkit/storage_tests.py [hunk 7]
```diff
@@ -86,12 +87,8 @@ def _create_test_threads(self, session, vm_ref_list):
         timeout function over SSH to every VM in vm_ref_list"""
         threads = []
         for vm_ref in vm_ref_list:
-            vm_ip = wait_for_ip(session, vm_ref, 'eth0')
-            threads.append(create_test_thread(lambda: TimeoutFunction(ssh_command(vm_ip,
-                                                                                  self.username,
-                                                                                  self.password,
-                                                                                  self.cmd_str),
-                                                                      self.timeout, '%s test timed out %d' % (self.test, self.timeout))))
+            threads.append(create_test_thread(ssh_command,
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/storage_tests.py]: Yes, will revert.

File: autocertkit/utils.py [hunk 8]
```diff
@@ -1103,74 +1103,193 @@ def _get_control_domain_ip(session, vm_ref, device='xenbr0'):
...
-    def should_timeout(start, timeout):
-        """Method for evaluating whether a time limit has been met"""
-        return time.time() - start > float(timeout)
 
-    i = 0
-    while not should_timeout(start, timeout):
-        log.debug("Trying to retrieve VM IP address - Attempt %d" % i)
-        ips = get_vm_ips(session, vm_ref)
-        log.debug("VM %s has these %s IPs" % (vm_ref, ips))
+def clean_context():
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Only different semantic that means init_context can be used once at beginning, but clean_context for other places.

File: autocertkit/utils.py [hunk 9]
```diff
@@ -1166,74 +1166,123 @@ def get_context_test_ifs(vm_ref):
     return [test_if for test_if in get_context_vm_ifs(vm_ref) if test_if[0] != mdevice]
 
 
-def wait_for_ip(session, vm_ref, device, timeout=300):
-    """Wait for an IP address to be returned (until a given timeout)"""
+def get_management_vif(session, vm_ref):
+    networks = get_management_network(session)
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Yes, it's a mistake. will fix

File: autocertkit/utils.py (hunk 1 above)
Comment (DeliZhangX) [reply autocertkit/utils.py]: No, I dislike textwrap. I will remove ```\```

File: autocertkit/utils.py [hunk 10]
```diff
@@ -1639,7 +1758,7 @@ def assert_can_boot_here(session, vm_ref, host_ref):
     return True
 
 
-def prepare_droid_vm(session, host_ref, creds=None):
+def prepare_droid_vm_template(session, host_ref, creds=None):
     """Checks if the droid vm needs to be installed
     on the host - if it does, it prepares it"""
     log.debug("About to prepare droid vm for host %s" % host_ref)
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Will change

File: autocertkit/utils.py [hunk 11]
```diff
@@ -1707,248 +1826,199 @@ def run_xapi_async_tasks(session, funcs, timeout=300):
     return [res for (_, res) in results]
 
 
-def deploy_count_droid_vms_on_host(session, host_ref, network_refs, vm_count, sms=None, sr_ref=None):
-    """Deploy vm_count VMs on the host_ref host. Required 
-    to define the network, optionally the SR"""
+def deploy_common_droid_vms_on_hosts(session, host_refs, network_refs, vm_count, sms=None, sr_ref=None):
+    """Deploy count common VMs on each host, here common means one network has only one vif,
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Will change

File: autocertkit/utils.py [hunk 12]
```diff
@@ -1707,248 +1826,199 @@ def run_xapi_async_tasks(session, funcs, timeout=300):
...
+        all_vms += vms
+        host_vm_list += [(host, vm) for vm in vms]
 
-    if vms:
-        # Our vm list should be empty if we have not timed out.
-        for vm in vms:
-            log.debug("VM not in '%s' state. Instead in '%s' state." %
-                      (power_state, session.xenapi.VM.get_power_state(vm)))
+    management_net_ref = network_refs[0] if len(network_refs) else None
+    test_net_ref = network_refs[1] if len(network_refs) > 1 else None
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: It's only follow before, network_refs[0] is management net, network_refs[1] is test net, at present no other test net.

File: autocertkit/utils.py [hunk 13]
```diff
@@ -1707,248 +1826,199 @@ def run_xapi_async_tasks(session, funcs, timeout=300):
...
+    return host_vms[master][0], host_vms[slave][0]
 
-    host_slave_ref = host_slave_refs[0]
 
+def import_droid_vms(session, host_ref, count=1, label="Droid"):
+    """Import count VM on host"""
     log.debug("Creating required VMs")
+    vm_template_ref = prepare_droid_vm_template(session, host_ref)
+    tasks = [lambda id=i: session.xenapi.Async.VM.clone(
+        vm_template_ref, '%s_%d' % (label, id)) for i in range(count)]
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Will change

File: autocertkit/utils.py [hunk 14]
```diff
@@ -1707,248 +1826,199 @@ def run_xapi_async_tasks(session, funcs, timeout=300):
...
+    log.debug("Setup vm %s vifs on network" % vm_ref)
+    vifs_rec = []
+    for id, vif_info in vifs_info.items():
+        vif_info = create_vif_on_vm_network(
+            session, vm_ref, network_ref, id, wipe=(id == 0), mac=vif_info[1])
+        vifs_rec.append(vif_info)
+    return vifs_rec
+
+
+def init_vifs_ip_addressing(session, vm_ref, vifs_info):
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Should not, they are different completely. If merged then you have to add more 'if else condition' branches

File: autocertkit/utils.py [hunk 15]
```diff
@@ -1707,248 +1826,199 @@ def run_xapi_async_tasks(session, funcs, timeout=300):
...
+    """Init VM vifs ip address by static or default dhcp"""
+    for id, vif_info in vifs_info.items():
+        device = "eth%d" % id
+        ip, netmask, gw = vif_info[2], vif_info[3], vif_info[4]
+        if ip:
+            log.debug("Init VM %s device %s as static IP: %s, netmask: %s, gateway: %s" % (
+                vm_ref, device, ip, netmask, gw))
+            droid_set_static(session, vm_ref, 'ipv4', device, ip, netmask, gw)
+        else:
+            log.debug("Init VM %s device %s as default dhcp" %
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Yes, dhcp is DEFAULT, no need parameter

File: autocertkit/utils.py [hunk 16]
```diff
@@ -1707,248 +1826,199 @@ def run_xapi_async_tasks(session, funcs, timeout=300):
...
 
-    return vif_list
+def init_droid_vm_vifs(session, vm_ref, network_ref, sms, ids=[1]):
+    """Init VM vifs"""
+    vm_vifs_info = alloc_vifs_info(session, vm_ref, network_ref, sms, ids)
+    init_ifs_ip_addressing(session, vm_ref, vm_vifs_info)
+    return vm_vifs_info
 
 
 def shutdown_droid_vms(session, vms, async=True):
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Will change

File: autocertkit/utils.py [hunk 17]
```diff
@@ -1992,89 +2062,46 @@ def start_droid_vms(session, vms, async=True):
...
-                             180)
-
-    except TimeoutFunctionException, e:
-        # Temporary ignore time out to start VM.
-        # If VM failed to start, test will fail while checking IPs.
-        log.debug("Timed out while starting VMs: %s" % e)
-        log.debug("Async call timed out but VM may started properly. tests go on.")
+    vm_refs = [vm_ref for _, vm_ref in vms]
+    # Check the VMs are in the 'Running' state.
+    wait_for_vms(session, vm_refs, XAPI_RUNNING_STATE)
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Will change start_droid_vms

File: autocertkit/utils.py [hunk 18]
```diff
@@ -1992,89 +2062,46 @@ def start_droid_vms(session, vms, async=True):
...
 
     # Temp fix for establishing that a VM has fully booted before
     # continuing with executing commands against it.
     log.debug("Wait for IPs...")
-    wait_for_all_ips(session, vm1_ref)
-    wait_for_all_ips(session, vm2_ref)
+    wait_for_vms_ips(session, vm_refs)
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Will change

File: autocertkit/utils.py [hunk 19]
```diff
@@ -2715,28 +2746,27 @@ def combine_recs(rec1, rec2):
     return rec
 
 
-def check_vm_ping_response(session, vm_ref, interface='eth0', count=3, timeout=300):
+def check_vm_ping_response(session, vm_ref, mip, interface='eth0', count=3, timeout=300):
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Will change

File: plugins/autocertkit [hunk 20]
```diff
@@ -292,17 +292,16 @@ def ssh_command(ip, username, password, cmd_str, dbg_str=None, attempts=10, time
     return {"returncode": -1, "stdout": "", "stderr": "An unkown error has occured!"}
 
 
-def install_ssh_key(session, vm_ref, username, password):
+def install_ssh_key(session, vm_ref, ip, username, password):
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply plugins/autocertkit]: This function is not top api, so no need to show the ip is management ip, instead it's a common ip as long as can ssh.

File: plugins/autocertkit [hunk 21]
```diff
@@ -501,54 +461,46 @@ def make_local_call(call, logging=True, std_out=subprocess.PIPE, shell=False):
 
 
 @log_exceptions
-def upload_file(session, vm_ref, username, password, file):
+def upload_file(session, vm_ref, ip, username, password, file):
     """upload one file to VM"""
-    ip_address = wait_for_ip(session, vm_ref, 'eth0')
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply plugins/autocertkit]: validate_exist is for plugin entry, upload_file is not plugin entry

File: plugins/autocertkit [hunk 22]
```diff
@@ -501,54 +461,46 @@ def make_local_call(call, logging=True, std_out=subprocess.PIPE, shell=False):
...
             "/opt/xensource/packages/files/auto-cert-kit/%s" % file,
-            'root@%s:' % ip_address]
+            'root@%s:' % ip]
     log.debug(call)
     make_local_call(call)
 
 
 @log_exceptions
-def run_command(session, vm_ref, username, password, cmd):
+def run_command(session, vm_ref, ip, username, password, cmd):
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply plugins/autocertkit]: You need to see original run_command where calls wait_for_ip().
I won't/shouldn't make change unrelated.

for utils.py autocerkit, you can think of they are two program can not call each other.

File: plugins/autocertkit [hunk 23]
```diff
@@ -501,54 +461,46 @@ def make_local_call(call, logging=True, std_out=subprocess.PIPE, shell=False):
...
-        upload_file(session, vm_ref, username, password, rpm)
+        upload_file(session, vm_ref, ip, username, password, rpm)
     rpm_paths = ["/root/%s" % i for i in rpms]
     cmd = "rpm -i %s --nosignature" % ' '.join(rpm_paths)
-    run_command(session, vm_ref, username, password, cmd)
+    run_command(session, vm_ref, ip, username, password, cmd)
 
 
-def install_rpm(session, vm_ref, username, password, rpm):
+def install_rpm(session, vm_ref, ip, username, password, rpm):
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply plugins/autocertkit]: I won't/shouldn't make change unrelated.

File: plugins/autocertkit [hunk 24]
```diff
@@ -501,54 +461,46 @@ def make_local_call(call, logging=True, std_out=subprocess.PIPE, shell=False):
...
-        install_rpm(session, vm_ref, username, password, package)
+        install_rpm(session, vm_ref, ip, username, password, package)
 
     if driver_name:
         cmd = "echo %s > /etc/modules-load.d/%s.conf" % (
             driver_name, driver_name)
-        run_command(session, vm_ref, username, password, cmd)
-
-
-def disable_net_dev_naming(session, vm_ref, username, password):
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply plugins/autocertkit]: Yes, disable_net_dev_naming is applied to any time, nothing to do ethx

File: plugins/autocertkit [hunk 25]
```diff
@@ -573,20 +525,20 @@ def deploy_iperf(session, args):
     """Install iPerf on a specified VM"""
     log.debug("deploy iperf plugin call")
     vm_ref = validate_exists(args, 'vm_ref')
-    username = validate_exists(args, 'username')
-    password = validate_exists(args, 'password')
 
     log.debug("validation complete")
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply plugins/autocertkit]: Will change

File: plugins/autocertkit (hunk 3 above)
Comment (DeliZhangX) [reply plugins/autocertkit]: VF driver won't install in dom0 at all

File: plugins/autocertkit [hunk 26]
```diff
@@ -631,43 +583,11 @@ def deploy_vf_driver(session, args):
     return json_dumps("OK")
 
 
-@log_exceptions
-def disable_network_device_naming(session, args):
-    """Disable network device_naming, referrence:
-        https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/networking_guide/ch-consistent_network_device_naming
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply plugins/autocertkit]: Yes, all interfaces naming by ifcfg. Including management interface, with binding the mac address

File: autocertkit/utils.py (hunk 2 above)
Comment (DeliZhangX) [reply autocertkit/utils.py]: I won't break the style, you can see in all sources, format() is not used.

File: autocertkit/utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (xihuan-citrix) [reply autocertkit/utils.py]: ` log.debug("About to prepare droid vm for host %s" % host_ref)` this still not modify

File: autocertkit/utils.py [hunk 27]
```diff
@@ -1103,74 +1103,193 @@ def _get_control_domain_ip(session, vm_ref, device='xenbr0'):
...
+
+def get_management_vif(session, vm_ref):
+    networks = get_management_network(session)
+    for vif in session.xenapi.VM.get_VIFs(vm_ref):
+        if session.xenapi.VIF.get_network(vif) in networks:
+            return vif
+    return None
+
+
+def init_vms_first_run(session, vm_refs):
```
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Will change

File: autocertkit/utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Changed

File: autocertkit/utils.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (DeliZhangX) [reply autocertkit/utils.py]: Changed

**Pull request merged**